cram_3d_world/cram_plan_transformation/src/predicates.lisp

```diff
@@ -0,0 +1,234 @@
+  ;;   (cpoe:task-specific-action ?top-level-name ?path :transporting-from-container ?compare ?_)
+  ;;   (not (== ?transport-action ?compare))
+  ;;   (without-replacement ?transport-action)
+  ;;   (without-replacement ?compare)
+  ;;   (location-distance-threshold ?threshold)
+  ;;   (task-nearby ?compare ?transport-action ?threshold :located-at)
+  ;;   (coe:task-full-path ?transport-action ?transport-path)
+  ;;   (cpoe:task-specific-action ?top-level-name ?transport-path :accessing-container ?access ?_)
+  ;;   (coe:task-full-path ?access ?accessing-path)
+  ;;   (cpoe:task-specific-action ?top-level-name ?accessing-path :navigating ?navigate ?navigate-action)
+  ;;   (cpoe:task-specific-action ?top-level-name ?transport-path :closing-container ?closing ?_)
+  ;;   (coe:task-full-path ?closing ?closing-path))
+  ;;; Transformation rule predicates ;;;
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+  #+unused-predicates-for-tray-transformation
+  ((<- (task-transporting-with-tray (?delivering-path))
+    (top-level-name ?top-level-name)
+    (top-level-path ?path)
+    (location-distance-threshold ?dist-threshold)
+    (cpoe:task-transporting-action ?top-level-name ?path ?last-task ?_)
+    (cpoe:task-transporting-action ?top-level-name ?path ?before-last-task ?_)
+    (not (== ?last-task ?before-last-task))
+    (without-replacement ?last-task)
+    (without-replacement ?before-last-task)
+    (task-location-description-equal ?last-task ?before-last-task)
+    (task-targets-nearby ?last-task ?before-last-task ?dist-threshold)
+    (coe:task-full-path ?last-task ?last-path)
+    (cpoe:task-delivering-action ?top-level-name ?last-path ?delivering-task ?_)
+    (coe:task-full-path ?delivering-task ?delivering-path))
+
+  (<- (task-transporting-with-tray-other-deliveries ?first-delivering-path (?other-path))
+    (bound ?first-delivering-path)
+    (top-level-name ?top-level-name)
+    (top-level-path ?path)
+    (location-distance-threshold ?dist-threshold)
+    (coe:task-from-path ?first-delivering-path ?del-task)
+    (cpoe:task-delivering-action ?top-level-name ?path ?other-del-task ?_)
+    (not (== ?other-del-task ?del-task))
+    (parent+ ?del-task ?task-transport)
+    (task-type ?task-transport :transporting)
+    (parent+ ?other-del-task ?other-task-transport)
+    (task-type ?other-task-transport :transporting)
+    (without-replacement ?task-transport) 
+    (without-replacement ?other-task-transport)
+    (task-location-description-equal ?task-transport ?other-task-transport)
+    (task-targets-nearby ?task-transport ?other-task-transport ?dist-threshold)
+    (task-full-path ?other-task-transport ?other-transp-path)
+    (task-delivering-action ?top-level-name ?other-transp-path ?other-delivering-task ?_)
+    (task-full-path ?other-delivering-task ?other-path))
+  
+  ;; Lazy parent backtracking. Should be able with subtask.
+  (<- (parent+ ?task ?parent)
```

Comment:
Go over all these predicates and see which ones can be put into cram_plan_occasions_events.